src/adaptation/precond.jl

```diff
@@ -2,389 +2,227 @@
+    if size(θ) != size(wv.var)
+        @assert wv.n == 0 "Cannot resize a var estimator when it contains samples."
+        wv.μ = zeros(T, size(θ))
+        wv.M = zeros(T, size(θ))
+        wv.δ = zeros(T, size(θ))
+        wv.var = ones(T, size(θ))
+    end
+end
 
-####
-#### Preconditioning matrix adaption implementation.
-####
+function reset!(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
+    wv.n = 0
+    wv.μ .= zero(T)
+    wv.M .= zero(T)
+end
 
-# Unit
-struct UnitPreconditioner{T} <: AbstractPreconditioner end
-Base.show(io::IO, ::UnitPreconditioner) = print(io, "UnitPreconditioner")
+function Base.push!(wv::WelfordVar, s::AbstractVecOrMat{T}) where {T}
+    wv.n += 1
+    @unpack δ, μ, M, n = wv
+    δ .= s - μ
+    μ .= μ + δ / n
+    M .= M + δ .* δ * (T(n - 1) / n)    # eqv. to `M + (s - μ) .* δ`
+end
 
-UnitPreconditioner(::Type{T}=Float64) where {T} = UnitPreconditioner{T}()
+# https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/var_adaptation.hpp
+function getest(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
+    @unpack n, M, var = wv
+    @assert n >= 2 "Cannot estimate variance with only one sample"
+    return T(n) / ((n + 5) * (n - 1)) * M .+ T(1e-3) * (5 / (n + 5))
+end
 
-string(::UnitPreconditioner) = "I"
-reset!(::UnitPreconditioner) = nothing
-getM⁻¹(dpc::UnitPreconditioner{T}) where {T<:AbstractFloat} = UniformScaling{T}(one(T))
-adapt!(
-    ::UnitPreconditioner,
-    ::AbstractVecOrMat{<:AbstractFloat},
-    ::AbstractScalarOrVec{<:AbstractFloat},
-    is_update::Bool=true
-) = nothing
+## Dense mass matrix adaptor
 
+abstract type CovEstimator{T} <: MassMatrixAdaptor end
 
-mutable struct DiagPreconditioner{
-    AT<:AbstractVecOrMat{<:AbstractFloat},
-    TEst<:VarEstimator{AT}
-} <: AbstractPreconditioner
-    n_min   :: Int
-    ve      :: TEst
-    var     :: AT
-end
-Base.show(io::IO, ::DiagPreconditioner) = print(io, "DiagPreconditioner")
-
-# Diagonal
-DiagPreconditioner(sz::Tuple{Vararg{Int}}, n_min::Int=10) = DiagPreconditioner(Float64, sz, n_min)
-function DiagPreconditioner(
-    ::Type{T},
-    sz::Tuple{Vararg{Int}};
-    n_min::Int=10,
-    var=ones(T, sz)
-) where {T}
-    ve = WelfordVar(T, sz)
-    return DiagPreconditioner(n_min, ve, var)
-end
+getM⁻¹(ce::CovEstimator) = ce.cov
 
-string(dpc::DiagPreconditioner) = string(dpc.var)
-reset!(dpc::DiagPreconditioner) = reset!(dpc.ve)
-getM⁻¹(dpc::DiagPreconditioner) = dpc.var
+Base.string(ce::CovEstimator) = string(LinearAlgebra.diag(getM⁻¹(ce)))
 
 function adapt!(
-    dpc::DiagPreconditioner,
+    ce::CovEstimator,
     θ::AbstractVecOrMat{<:AbstractFloat},
     α::AbstractScalarOrVec{<:AbstractFloat},
     is_update::Bool=true
 )
-    resize!(dpc, θ)
-    add_sample!(dpc.ve, θ)
-    if dpc.ve.n >= dpc.n_min && is_update
-        # Can be made inplace
-        dpc.var .= get_var(dpc.ve)
+    resize!(ce, θ)
+    push!(ce, θ)
+    if ce.n >= ce.n_min && is_update
+        ce.cov .= getest(ce)
```

Comment:
It seems it's basically the same logic for the variance and the covariance estimators - maybe just define a function `update!(estimator)` that updates the estimators, so in one case `ve.var` and in the other case `ce.cov`. Then one could just have one `adapt!` method for both of them.